### ISSUE TITLE
[FEATURE REQ] DefaultAzureCredential should use "Try" methods instead of relying on Exceptions

### ISSUE THREAD

**Issue by peter-bertok** [Azure/azure-sdk-for-net#27040]
### Library name

Azure.Identity

### Please describe the feature.

This is related to the block of code here: https://github.com/Azure/azure-sdk-for-net/blob/4b2579556b7271587d2fb122163e23090a043597/sdk/identity/Azure.Identity/src/DefaultAzureCredential.cs#L167-L181

The `GetTokenFromSourcesAsync()` and related functions sequentially try to acquire credentials from a list of `TokenCredential` sources. The problem with this is that `DefaultAzureCredential` is heavily used in many SDKs, and the list of sources is long. This generates enormous "exception spam" which pollutes logs such as Application Insights, debug output, etc...

Throwing exceptions like this is very inefficient as well, noticeably slowing down clients that use the Azure SDK, _even if_ they don't directly use this class!

1) To cater for scenarios like this where *failure is expected*, methods such as `TryGetTokenAsync` should be added to the `TokenCredential` that return a boolean "false" on failure instead of throwing an exception. If the error message is desired, then that should be returned via an `out` parameter (or some sort of Diagnostics context) instead of an `Exception`.

2) Consider creating a class that is a generic list of token sources that can be tried in order. The `DefaultAzureCredential` class is not the only scenario where this makes sense. Instead of a fixed-length array of 8 items maximum, use a List<TokenCredential>. This new general-purpose class could then be used by `DefaultAzureCredential` _and also_ by Microsoft customers that use this SDK.

3) The order in which credentials are tried doesn't particularly matter, as long as the first one that succeeds in the sequence originally specified is selected. A significant performance boost may be possible by trying all sources in parallel via `Task.WaitAny()`.

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team members best able to assist.

**Comment by christothes** [comment]
Hi @peter-bertok - Thanks for the feedback.

> 1. To cater for scenarios like this where _failure is expected_, methods such as `TryGetTokenAsync` should be added to the `TokenCredential` that return a boolean "false" on failure instead of throwing an exception. If the error message is desired, then that should be returned via an `out` parameter (or some sort of Diagnostics context) instead of an `Exception`.

Regarding usage of exceptions to indicate error scenarios, this is consistent with both the .NET and Azure SDK design guidelines. See the below links for reference.
https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/exception-throwing
https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-errors

In addition, out parameters are not supported in async methods, so a TryGetTokenAsync pattern would not work well here.

> 2. Consider creating a class that is a generic list of token sources that can be tried in order. The `DefaultAzureCredential` class is not the only scenario where this makes sense. Instead of a fixed-length array of 8 items maximum, use a List. This new general-purpose class could then be used by `DefaultAzureCredential` _and also_ by Microsoft customers that use this SDK.

We have this today - it's called `ChainedTokenCredential` - https://github.com/Azure/azure-sdk-for-net/blob/32061eceda56d8a86381c3a4a309bc0ba7aa8ec7/sdk/identity/Azure.Identity/README.md#define-a-custom-authentication-flow-with-the-chainedtokencredential

> 3. The order in which credentials are tried doesn't particularly matter, as long as the first one that succeeds in the sequence originally specified is selected. A significant performance boost may be possible by trying all sources in parallel via `Task.WaitAny()`.

If ordering is to be preserved on selection, then you'd always be bound by the response time of the earlier credentials in the chain. I think this would defeat the purpose of a `WaitAny()` call. In addition, for production scenarios where performance matters most, the latter credentials in the chain are typically never used or considered.

In summary, `DefaultAzureCredential` is designed to be the best "getting started" credential choice and is intended to satisfy the most common scenarios. If performance optimization is a goal, I'd recommend configuring the `DefaultAzureCredentialOptions` to eliminate unwanted credentials from the chain or constructing a `ChainedTokenCredential` containing only the credentials required for your scenario, ordered for optimal performance. In cases where controlling the exact behavior is required, it is also possible to create a custom implementation of `TokenCredential`.

**Comment by peter-bertok** [comment]
> Regarding usage of exceptions to indicate error scenarios, this is consistent with both the .NET and Azure SDK design guidelines. See the below links for reference. https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/exception-throwing https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-errors

High-level rules for user-facing API surfaces do not necessarily apply to internal framework logic. These calls are _expected_ to fail, which by definition makes them _non-exceptional_. A standard optimization is to use "Try" methods, or some other method of detecting failure without throwing an exception. This is common practice throughout the .NET standard library.

> In addition, out parameters are not supported in async methods, so a TryGetTokenAsync pattern would not work well here.

Sure, but it's possible to return a class or tuple type with multiple values in it, including a success/failure code and error message. An exception should be thrown only if _all_ authentication sources fail.

> If ordering is to be preserved on selection, then you'd always be bound by the response time of the earlier credentials in the chain. I think this would defeat the purpose of a `WaitAny()` call. In addition, for production scenarios where performance matters most, the latter credentials in the chain are typically never used or considered.

It's surprisingly common to run code on highly locked-down Azure servers, developer workstations with web proxies, disconnected Internet, etc...

The performance of these calls matters because they're in the critical path for many SDK components.

If any of them reach out to the network, e.g.: 169.254.169.254 or the AAD endpoints, then these should be overlapped to the maximum extent possible to minimize the latency hit if there's a timeout. A rough outline of the code would be:

```csharp
var tasks = new Task<TryAuthResult>[8];
for( ... ) { tasks[i] = sources[i].TryAuthAsync(...); }
for( ... ) { if (( await tasks[i]).Success return tasks[i].Result;  }
// If nothing short-circuited above, throw aggregate exception here
// Cancel remaining tasks in a finally block.
```

> In summary, `DefaultAzureCredential` is designed to be the best "getting started" credential choice and is intended to satisfy the most common scenarios. If performance optimization is a goal, I'd recommend configuring the `DefaultAzureCredentialOptions` to eliminate unwanted credentials from the chain or constructing a `ChainedTokenCredential` containing only the credentials required for your scenario, ordered for optimal performance. In cases where controlling the exact behavior is required, it is also possible to create a custom implementation of `TokenCredential`.

In my humble opinion, performance should be **everyone's goal, always**. Aiming for the highest possible quality is especially important for such commonly used libraries. 

Simply including a couple of Azure SDK NuGet packages has seen my web app startup time go from milliseconds to tens of seconds! Watching the logs shows that an awful lot of time is spent on internal "noise" like this. It's distressing to see Microsoft's .NET slow down to Enterprise Java levels.

As a random example, the Application Insights SDK pulls in Azure.Identity and invokes this class without any code written by the end-user! Customizing its behavior for performance is not always an option, because it has a "codeless attach" mode.

Please consider all end-users, not just demo or "getting started" scenarios.

**Comment by christothes** [comment]
>  Simply including a couple of Azure SDK NuGet packages has seen my web app startup time go from milliseconds to tens of seconds! Watching the logs shows that an awful lot of time is spent on internal "noise" like this. It's distressing to see Microsoft's .NET slow down to Enterprise Java levels.

To clarify - we definitely value performance in the SDKs. However, I'd be quite surprised if the the exceptions being thrown is the significant factor in startup times you are experiencing. 

We are aware of a recent performance regression in 1.5.0 related to the ManagedIdentity endpoint discovery in DefaultAzureCredential (see #24767), so I'm curious if this is what you are experiencing. Setting the [ExcludeManagedIdentityCredential](https://github.com/Azure/azure-sdk-for-net/blob/67ac8347f3ed5c0c617f5d3c39debac4afb8d655/sdk/identity/Azure.Identity/src/DefaultAzureCredentialOptions.cs#L80) property in the `DefaultAzureCredentialOptions` would confirm.
